### PR TITLE
Make transaction commit async

### PR DESCRIPTION
**Pull request by JPDSousa**
This commit pushes most of the commit operation to a dedicated single-threaded pool. Thus, commits are async from the calling thread, but still serializable between each other.

### CODE REVIEW

**Comment by JPDSousa** [comment]
@JohannesLichtenberger there's still a lot of uncertainty from my side. I wanted you to take a look at this to check if we're going on the right direction. If so, then there's some boilerplate tasks that need to be performed before this MR gets closed:
* Tests
* Documentation
* Some further refactoring

**Comment by JohannesLichtenberger** [comment]
Hi, I'm not sure how you handle the async commit stuff, that is when then transaction is committed we can fill a new in-memory transaction intent log buffer already and also parse XML or JSON :-) that is we need two transaction logs, I guess... one that is bound to the committing transaction and another one, which already fills another transaction intent log. I think the problem we face is, that the transaction is still open even after committing. Plus no two threads are allowed to commit at the same time, so it has to block once the next commit is going to be executed and the first one still commits. I have some changes locally, but never got it running so far, because we also need new UberPages and RevisionRootPages.

I think Sebastian got it running (I've forked the university project some time before the changes happened):

https://github.com/sebastiangraf/treetank/blob/master/coremodules/core/src/main/java/org/treetank/access/BucketWriteTrx.java

I didn't like the direction the project took, especially regarding API changes... but it's also dead since he left the university.

I think we can't implement it in the `PageTrx`, though and have to implement it in the `JsonNodeTrx` and `XmlNodeTrx`. But maybe I missed something and your proposed changes already work, despite the failing unit or integration tests.

kind regards
Johannes

**Review by JPDSousa — COMMENTED** [review]
Hi @JohannesLichtenberger 

Sorry for the late reply!

> Hi, I'm not sure how you handle the async commit stuff

Sorry. That's totally my fault. I started changing parts of the code that were not directly related with the async-commit patch, and that made this MR quite messy. I may be able to split this into multiple MRs if that reveals itself a problem as you review this.

In regards to the actual solution, I'm handling that in `AbstractNodeTrxImpl#commit` ([here](https://github.com/sirixdb/sirix/pull/410/files#diff-c678b6d6a4988b7bc5b53f79ae4c8b0ccbd754b851e5079b43d20e9b4ce89999R293)). I created `TransactionCommiter` which is responsible for handling the commit logic asynchronously. The way I thought about it (which I might change in the meantime) is that we pass the `pageTrx` and everything else we want to commit and it will perform the actual commit logic in a separate thread. `PageTrx#commit` is also called inside this thread, so it's considered async work. In the meatime, the rest of `AbstractNodeTrxImpl#commit` creates a new root page and works from there.

> that is when then transaction is committed we can fill a new in-memory transaction intent log buffer already and also parse XML or JSON :-) that is we need two transaction logs, I guess... one that is bound to the committing transaction and another one, which already fills another transaction intent log. I think the problem we face is, that the transaction is still open even after committing. 

I believe I understand your point here. It looks like it goes towards my implementation, but since I'm still learning a lot about the transaction lifecycle and architecture, I might be missing something. 

Note that this approach is optimistic, in the sense that it assumes the commit is successful. However, I don't have a simple solution to avoid this optimistic approach.

> Plus no two threads are allowed to commit at the same time, so it has to block once the next commit is going to be executed and the first one still commits. 

This part is guaranteed by the fact that the "committer pool" in `TransactionCommiter` is single threaded, for both recurrent commits and "normal" commits.

>I have some changes locally, but never got it running so far, because we also need new UberPages and RevisionRootPages.

Can you push them to a branch so that I can look at them and try to figure out possible improvements on this diff?

> I think Sebastian got it running (I've forked the university project some time before the changes happened):
> 
> https://github.com/sebastiangraf/treetank/blob/master/coremodules/core/src/main/java/org/treetank/access/BucketWriteTrx.java
> 
> I didn't like the direction the project took, especially regarding API changes... but it's also dead since he left the university.

I'll take a look at this to try to figure out possible improvements as well.

> I think we can't implement it in the `PageTrx`, though and have to implement it in the `JsonNodeTrx` and `XmlNodeTrx`. But maybe I missed something and your proposed changes already work, despite the failing unit or integration tests.

From what I understand a commit in a transaction will always trigger a commit in the page. If so, then if we move the "commit logic" into a dedicated thread then the page "commit logic" will implicitly be moved as well. It looks like my understanding matches your explanation, and also matches my (intended) implementation here. 

Having said this, I think I have some action points on my side:
* Look at Sebastian's code to gather possible improvements to this MR
* Implement some improvements to `TransactionCommitter` that'll make the code more cohesive.

I think that besides that I'll wait for your re-review to gather additional next steps.

**Comment by JohannesLichtenberger** [comment]
> Hi @JohannesLichtenberger
> 
> Sorry for the late reply!
> 
> > Hi, I'm not sure how you handle the async commit stuff
> 
> Sorry. That's totally my fault. I started changing parts of the code that were not directly related with the async-commit patch, and that made this MR quite messy. I may be able to split this into multiple MRs if that reveals itself a problem as you review this.
> 
> In regards to the actual solution, I'm handling that in `AbstractNodeTrxImpl#commit` ([here](https://github.com/sirixdb/sirix/pull/410/files#diff-c678b6d6a4988b7bc5b53f79ae4c8b0ccbd754b851e5079b43d20e9b4ce89999R293)). I created `TransactionCommiter` which is responsible for handling the commit logic asynchronously. The way I thought about it (which I might change in the meantime) is that we pass the `pageTrx` and everything else we want to commit and it will perform the actual commit logic in a separate thread. `PageTrx#commit` is also called inside this thread, so it's considered async work. In the meatime, the rest of `AbstractNodeTrxImpl#commit` creates a new root page and works from there.

The problem with this approach is, that the new `UberPage` and `RevisionRootPage` must be based on the previous one (which is about to be written and not registered in the ResourceManager at the time a new `NodePageTrx` is started. Thus, ResourceManager at the point of time does not know, that it must be maxRevision + 2, at the time a new node page trx is started. Furthermore, the page trx has to receive the former trx intent log, to read from this log, if it can't find the page in-memory or on disk (so if it's about to be written in the commit thread).

So, I think it's rather tricky to get everything correctly implemented. I think the main use case will be a longer running initial import of a huge JSON file, which can hopefully significantly improve the import-performance, if we already sync a trx intent log to disk and on the other hand fill a new trx intent log with parse XML/JSON nodes.

I'll try to push my changes into my sirix branch. However, I haven't integrated all your changes as of now...

I'm usually checking the changes importing a huge chicago.json file (about 3,8 Gb) and setting a time based auto-commit every few seconds. However, I think in that case also the transaction intent log has to be synchronized, such that on the one hand the in-memory log can be invalidated (clearing the map basically), and on the other hand another thread can read from the cache in parallel.

So, I think your approach as of now doesn't tackle the issue in particular, that the `UberPage` and `RevisionRootPage` are still based on the first max revision, when the long running read-write transaction was started and as such it's not based on the currently about to be written pages.

File: bundles/sirix-core/src/main/java/org/sirix/access/trx/node/AbstractNodeTrxImpl.java [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+                reInstantiate(getId(), revisionNumber);
+                state = State.Running;
+            } else {
+                state = State.Committed;
+            }
+        });
+
+        // Execute post-commit hooks.
+        // TODO should commit hooks execute immediately?
+        for (final PostCommitHook hook : postCommitHooks) {
```
[the remark this answers is not in the document]
Comment (JohannesLichtenberger) [reply bundles/sirix-core/src/main/java/org/sirix/access/trx/node/AbstractNodeTrxImpl.java:315]: Yes, that's true. That's basically as of now luckily not used ;-)

**Comment by JohannesLichtenberger** [comment]
I've pushed my async work I've done some time ago to https://github.com/JohannesLichtenberger/sirix

However, I know the method signature is stupid, the transaction intent log has to be synchronized and it didn't work with the Chicago (historical data since 2001) dataset from https://github.com/jdorfman/awesome-json-datasets

The integration test I created is in `JsonShredderTest.testChicago()`. It basically should commit every 3 seconds (that is the scheduled thread should sync to the persistent storage in the background), while a new page trx is created (`JsonNodeTrxImpl`) and the JSON file is parsed and new in-memory nodes are gathered in a new transaction intent log, while the old (currently in committing state stuff might be read from the old transaction intent log). Some committed nodes, however couldn't be found and the test fails after the first commit, basically.

**Comment by JohannesLichtenberger** [comment]
Beware that the file I'm using is 3,9 Gb large, I think :)

**Comment by JohannesLichtenberger** [comment]
I hope this helps a little bit. Let me know if you need more infos :+1:

**Comment by JPDSousa** [comment]
Hey @JohannesLichtenberger 

Sorry for my absence. Long story short I'm currently in between jobs, and I'm going to be travelling for the next month. I'm sorry about this, but it was a sudden decision, and among all the things I had to change, this PR got thrown to the backlog.

Feel free to comandeer this any way you like. I've read the comments and I now realize that this implementation diverges from what was intended. Last case scenario, feel free to close this, and I'll take care of extracting all the unrelated refactoring into separate PRs.

I believe I'll be available to contributing to the project somewhere during august.

**Comment by JohannesLichtenberger** [comment]
Sounds great, maybe you like to keep working on this stuff in August/September? I've already spent some time and it seems it's not that simple... so, I'm pushing this always back currently and it somehow stopped me from working on the project, as I became a bit frustrated about this stuff (despite, I got a new kitchen and had to renovate the room). But in fact, I'm currently always postponing my work on SirixDB, because of this, as I think it's the most pressing issue (despite a new index-tree based on a HOT (height optimized trie)).

**Comment by JohannesLichtenberger** [comment]
Once we've implemented this and once we have a working frontend to compare revisions... I think we should make another post on Hacker News :+1: However, I know that we are too few people working on the frontend to get a production-ready version, I guess. Despite, I realized that byte-addressable persistent memory could be key for efficiency (fetching a number of small page-fragements efficiently -- a predefined number currently is used for a sliding window to reconstruct an in-memory page). Maybe we don't even always need the whole page in-memory. There's work from Viktor Leis with a small in-memory buffer of "mini-pages"...

**Comment by JohannesLichtenberger** [comment]
Nice travels :-)

**Comment by JohannesLichtenberger** [comment]
@JPDSousa hope you had great travels and that you're going to like your new job :+1: 

Are you going to work on this issue? I think it would be best to also implement the async commit in the page read-write transaction insteaf of on the higher level (`JsonNodeTrx` / `XmlNodeTrx`) as in Treetank.

**Comment by JPDSousa** [comment]
> @JPDSousa hope you had great travels and that you're going to like your new job 👍
> 
> Are you going to work on this issue? I think it would be best to also implement the async commit in the page read-write transaction insteaf of on the higher level (`JsonNodeTrx` / `XmlNodeTrx`). Just as in Treetank where Sebastian renamed the trx to the `BucketWriteTrx`.

Hey @JohannesLichtenberger ! Thank you for remembering! I did have great travels, and my new job looks amazing 👍 

In regards to this thread. I definitely wanna get back to it. When I last touched this, I had some questions regarding the asynchrony of the commits. i.e., We should have a dedicated thread that runs all sync, async and automatic commits, right? Besides that I'm still a bit confused about the way we handle failing commits (I see this async commit approach as an optimistic, if we do not provide a hook/callback where callers can check whether the commit failed or not. Also, I still need to wrap my head around all the classes that need to be changed. On this last point, as far as I remember, you already mentioned a list of classes and patches I should look as starting points.

In terms of planning, I guess my first quick step is going to be extracting all the refactors in this PR to a separate one, to get some clarity. I know that this looks like a "noop" in terms of async commit, but it will not only help me separate the changes that will get landed anyway from the ones I need to change according to your suggestions.

Anyway, feel free to step in at any point to take over (some or all) this initiative. I expect to pick in up in the next days, whenever I can find a couple hours to get back on track.

**Comment by JohannesLichtenberger** [comment]
My idea of the async commit stuff first and foremost is, that we can shred/import data more quickly. That is it's some sort of an optimistic commit. If we use the scheduler executor service to for instance to trigger a commit every 3 or 4 seconds the commit itself / the sync to persistent storage could be done in parallel to insertions of new nodes into the in-memory trx intent log / the simple buffer, thus it would be some kind of NIO (plus the overhead of serializing the pages to persistent storage in a postorder traversal).

We don't want to wait for the commit -- the sync to a flash drive to finish. We however should make sure, that only one commit is triggered through a simple Lock in the background. Thus, at most we have one commit which is broken while we ingest data into SirixDB. If we detect, for instance a .commit file is still in the resource directory the next time a read-write trx is started we should simply truncate the storage log to the last committed UberPage (which is the main root-node into the huge page-tree representing a resource). The UberPages are currently stored in a backwards linked list (every UberPage has a link to it's predecessor), but we'd probably do not need this and could also simply store the current UberPage plus the previous UberPage and replicate it a few times as in ZFS (ZFS was a huge inspiration).

I think what you want and maybe that's even orthogonal is a process or thread which gets the transactions or the trx intent logs to commit and triggers a commit after some data has been gathered or a certain timespan is over. That way small transactions from different resources in a database can be gathered first, making the sync faster (instead of syncing a lot of small transactions each on it's own to persistent storage). This would be in-line with how Datomic approaches the trx commits (the Transactor) and it reminds me of Caetano Sauers work (I think SirixDB is pretty much in-line with most of the stuff described theoretically here, as the resource itself is a huge indexed log and we don't need a second, persistent structure, the trx log -- only an in-memory buffer): http://wwwlgis.informatik.uni-kl.de/cms/fileadmin/publications/2018/PVLDB2018-Sauer.pdf

**Comment by JohannesLichtenberger** [comment]
I think you might be able to use my branch https://github.com/JohannesLichtenberger/sirix as an inspiration for the first approach (for now without the thread or process which commits a bunch of trx in-memory logs together). However, I didn't merge in all the stuff we changed, lately.

**Comment by JPDSousa** [comment]
Closing this, as we've extracted all the value in #429 and #428

**Pull request closed without merging**